### PR TITLE
Added more recoding options into recode numeric dialog

### LINKED ISSUE #6851: Improve the Prepare > Factor > Recode Numeric dialogue
This is a relatively simple task.  Could @shadrackkibet suggest who could have the time to do this.

This dialogue currently uses the default cut function.  I propose instead we use the `Hmisc::cut2` function instead.
Here is the current dialogue:
image

1) We no longer need the `Closed on` group box, 
2) There are 3 radio buttons under the single receiver.  Possibly in a group box.  The 3rd is the Existing Break Points as it is now
3) The first is labelled Quantiles and is followed, if chosen, by an up-down with minimum 1 and no maximum, and default say 4.
4) the middle one is labelled Minimum and again has an up-down with miit includes nimum 1 and no maximum.
5) Change the Add Labels checkbox to `Label Groups with Means`.  Get rid of the field to type in.  If chosen the command adds levels.mean = TRUE and also digits=3

### PR DESCRIPTION
**Pull request by mmulahvictor**
Fixes #6851
@africanmathsinitiative/developers this is ready for review
@rdstern @shadrackkibet have a look at it.

### CODE REVIEW

**Comment by mmulahvictor** [comment]
@rdstern I have completed the changes above.

**Review by rdstern — DISMISSED** [review]
@mmulahvictor great.  I am approving.  Let's see what @lloyddewit or @shadrackkibet feel.

**Comment by mmulahvictor** [comment]
@N-thony I have made the changes

**Comment by N-thony** [comment]
@mmulahvictor thanks. Over to you @shadrackkibet

**Comment by shadrackkibet** [comment]
Very minor suggestions.

- Push the receiver label to left a bit... so that it aligns properly
- Should Ok be enabled in this kind of instance?
image
image
image

- The **Label Groups with Means** doesn't reset?
- Change the save control label to **New Column Name:** also **Break Points:**?
- Perhaps, a general problem with the save control. Reset doesn't work when you switch position. I claim it should switch to the default.

File: instat/dlgRecodeNumeric.vb [hunk 1]
```diff
@@ -80,86 +97,44 @@ Public Class dlgRecodeNumeric
...
-        clsCutFunction.AddParameter("include.lowest", "TRUE", iPosition:=0)
-        clsCutFunction.AddParameter("dig.lab", "10", iPosition:=1)
-        clsCutFunction.AddParameter("right", "FALSE", iPosition:=3)
+        clsDummyFunction.AddParameter("checked", "quantile", iPosition:=0)
 
-        ucrBase.clsRsyntax.SetBaseRFunction(clsCutFunction)
+        clsCut2Function.SetRCommand("cut2")
+        clsCut2Function.SetPackageName("Hmisc")
+
+        AddParameters()
```
Comment (shadrackkibet) [inline instat/dlgRecodeNumeric.vb:113]: Could you explain why you added this here?

**Review by shadrackkibet — CHANGES_REQUESTED** [review]

**Comment by rdstern** [comment]
In addition:
a) The group box is labelled Quantile Groups.  It should be just Groups.
b) I tried with the gss_cat data from the forcats package.  I started with the 3rd button and the limits 30,50,70 and it worked fine.  Then I changed to the quantiles, and it still repeated the previous command.
c) I then did reset and tried the quantiles again.  When I use the default of 4 it gives the command as 
`recode <- Hmisc::cut2(x=age) ` which is not what's wanted!  
If I change to 5 it is ok, i.e. `recode2 <- Hmisc::cut2(g=5, x=age)`  so I think you just need to include g=4 above.
The option with m=2000 worked ok, please check also with the default, that it is also fine!

**Comment by mmulahvictor** [comment]
@shadrackkibet I have completed the changes above.

**Comment by shadrackkibet** [comment]
@mmulahvictor I did refactor the code a bit. Pull the changes and test.

**Comment by mmulahvictor** [comment]
@shadrackkibet When I select the Quantiles radio button and do aReset, it brings the following error.
Screenshot (26)

**Comment by lloyddewit** [comment]
@rdstern if you can test/approve again, then I'll do the final peer review and (hopefully) merge.
thanks

**Review by rdstern — CHANGES_REQUESTED** [review]
This is looking good now.
a) Could ypou make 4 the default for the quantiles, and (say) 5 for the Minimum.  (They are currently 1.)
b) The means option should be unticked by default.  It doesn't yet work.  I don't see a change in the code when it is ticked.

**Comment by mmulahvictor** [comment]
@rdstern and @shadrackkibet can you look at this change. This is a screenshot of the developer error.
Screenshot (32)

**Comment by shadrackkibet** [comment]
@mmulahvictor anything remaining? I can help sort it out now.

**Comment by mmulahvictor** [comment]
@shadrackkibet  everything is done now.

File: instat/dlgRecodeNumeric.vb [hunk 2]
```diff
@@ -53,10 +53,10 @@ Public Class dlgRecodeNumeric
         ucrReceiverRecode.SetMeAsReceiver()
 
         ucrNudQuantiles.SetParameter(New RParameter("g", 1))
-        ucrNudQuantiles.SetMinMax(1)
+        ucrNudQuantiles.SetMinMax(4)
 
         ucrNudMinimum.SetParameter(New RParameter("m", 2))
-        ucrNudMinimum.SetMinMax(1)
+        ucrNudMinimum.SetMinMax(5)
```
Comment (shadrackkibet) [inline instat/dlgRecodeNumeric.vb]: I think it is correct to set the minimum value as 1. We want `4` and `5` as the parameter defaults and not the minimum value a control can allow. This means the user can't input a value less than 4 or 5 respectively. To sort this out add these as parameter vaues in setDefaults.

**Review by shadrackkibet — CHANGES_REQUESTED** [review]

**Comment by lloyddewit** [comment]
@mmulahvictor Thank you for this PR. 
Do you believe that you've now resolved all the comments from @rdstern and @shadrackkibet? 
Is this PR ready for final approvals and merge?
Thanks for clarifying

**Comment by mmulahvictor** [comment]
Thank you @lloyddewit 
I have resolved all the changes.

**Comment by lloyddewit** [comment]
@rdstern @shadrackkibet Please could you give your final approvals? thanks

**Comment by shadrackkibet** [comment]
@mmulahvictor please update this branch with the master branch.

**Review by rdstern — CHANGES_REQUESTED** [review]
The default for quantiles is still 1.  Could it be made (say) 4.  And could the word quantiles be changed to become Levels.  It is just the number of levels of the factor being produced.
It is looking good now, so I would like to get it merged.

**Comment by mmulahvictor** [comment]
> The default for quantiles is still 1. Could it be made (say) 4. And could the word quantiles be changed to become Levels. It is just the number of levels of the factor being produced. It is looking good now, so I would like to get it merged.

@rdstern I made the changes on the default for the Quantiles and it is working. The images below shows that the least you can select is 1 and the default is 4 respectively.

Screenshot (35)
Screenshot (34)

**Comment by lloyddewit** [comment]
@rdstern @shadrackkibet Please could you give your final approvals? thanks

**Review by rdstern — DISMISSED** [review]
Seems fine now

**Comment by shadrackkibet** [comment]
@mmulahvictor could you sort out the conflict then we can merge?

**Comment by mmulahvictor** [comment]
@shadrackkibet  I have resolved the issue

**Comment by shadrackkibet** [comment]
@mmulahvictor I can't test it. I get the following build errors.
image

**Comment by shadrackkibet** [comment]
I get more build errors. Looks like there is a problem with this file. Perhaps that is because you manually edited the resource file which is normally very tricky.

image

The simplest way to sort this out is to redo these changes in a new PR. You could just copy-paste carefully.

**Comment by shadrackkibet** [comment]
@mmulahvictor any progress on this?

**Comment by mmulahvictor** [comment]
> @mmulahvictor any progress on this?

I have a system crash right now which I am trying to restore it but I only remain with changes on the designer.

**Comment by shadrackkibet** [comment]
I hope the crash was sorted out? Could you report on the progress here?

**Comment by mmulahvictor** [comment]
Yeah it is sorted out. I am currently installing R-Instat on Ubuntu as suggested by Maxwell and Danny. I will complete this by today and continue with this work afterwards.

**Comment by lloyddewit** [comment]
replaced by PR #7136

**Pull request closed without merging**